File: src/parallel_region.f90
PROGRAM parallel_region
#ifdef OPENMP
	USE OMP_LIB
#endif
	IMPLICIT NONE
	INTEGER(kind=4)			:: length
	REAL(kind=8)			:: tmp, res
	length = 1024*1024
	call omp_subroutine(length,res)
	write(*,100) "The result of (arr(1)+arr(",length,") is ", res
	100 format (A,I7,A,e13.6e2)
CONTAINS
	SUBROUTINE omp_subroutine(length,res)
		IMPLICIT NONE
		INTEGER(kind=4), INTENT(in)	:: length
		INTEGER(kind=4)			:: i,tid,threads,from,to,block_size
		REAL(kind=8), ALLOCATABLE	:: arr(:)
		REAL(kind=8)			:: tmp
		REAL(kind=8), INTENT(out)	:: res 
		allocate (arr(length))
		!$omp parallel private(tid,threads,i,from,to,block_size) shared(length,arr)
#ifdef OPENMP
		tid = omp_get_thread_num()
		threads = omp_get_num_threads()
#else
		tid = 0
		threads = 1
#endif
		block_size = length / threads
		from = tid*block_size + 1 
		IF (tid .EQ. (threads - 1)) THEN
			to = length
		ELSE
			to = from + block_size
		END IF
		DO i=from,to
			arr(i) = 1.0/length
	        	tmp = arr(i)
		END DO
		!$omp end parallel
		res = arr(1) + arr(length)
		deallocate(arr)
	END SUBROUTINE omp_subroutine
END PROGRAM parallel_region

File: src/parallel_region.c
#include <stdio.h>
#include <stdlib.h>
#ifdef OPENMP
#include <omp.h>
#endif
int main(void) {
	const int length = 1024*1024;
	double * arr = (double *) malloc(length*sizeof(double));
	#pragma omp parallel
	{
#ifdef OPENMP
		int tid = omp_get_thread_num();
		int threads = omp_get_num_threads();
#else
		int tid = 0;
		int threads = 1;
#endif
		int block_size = length/threads;
		int from = tid*block_size;
		int to = (tid == threads-1) ? length : from+block_size; 
		for(int i = from; i < to; i++) {
			arr[i] = 1.0/length;
		}
	}
	printf("The result of arr[0]+arr[%d] is %lf\n",length-1,arr[0]+arr[length-1]);
	free(arr);
	return 0;
}
